File: frontend/src/app/core/services/announcement.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface NSEAnnouncement {
  symbol: string;
  company_name: string;
  subject: string;
  broadcast_date: string;
  attachment_link: string;
  category: string;
}

export interface BSEAnnouncement {
  scrip_code: string;
  company_name: string;
  subject: string;
  news_date: string;
  category: string;
  attachment_url: string | null;
  news_id: string;
}

export interface NSEAnnouncementsResponse {
  symbol: string;
  count: number;
  announcements: NSEAnnouncement[];
  message?: string;
}

export interface BSEAnnouncementsResponse {
  scrip_code: string | null;
  from_date: string | null;
  to_date: string | null;
  announcements: BSEAnnouncement[];
  total_pages: number;
  current_page: number;
}

@Injectable({
  providedIn: 'root'
})
export class AnnouncementService {
  private apiUrl = `${import.meta.env.NG_APP_BACKEND}/api/announcements`;

  constructor(private http: HttpClient) { }

  getNSEAnnouncements(symbol: string, fromDate?: string, toDate?: string, limit: number = 100): Observable<NSEAnnouncementsResponse> {
    let params = new HttpParams().set('limit', limit.toString());

    if (fromDate) {
      params = params.set('from_date', fromDate);
    }
    if (toDate) {
      params = params.set('to_date', toDate);
    }

    return this.http.get<NSEAnnouncementsResponse>(`${this.apiUrl}/nse/${symbol}`, { params });
  }

  getBSEAnnouncements(
    scripCode?: string,
    fromDate?: string,
    toDate?: string,
    page: number = 1
  ): Observable<BSEAnnouncementsResponse> {
    let params = new HttpParams().set('page', page.toString());

    if (scripCode) {
      params = params.set('scrip_code', scripCode);
    }
    if (fromDate) {
      params = params.set('from_date', fromDate);
    }
    if (toDate) {
      params = params.set('to_date', toDate);
    }

    return this.http.get<BSEAnnouncementsResponse>(`${this.apiUrl}/bse`, { params });
  }

  getBSEScripCodes(): Observable<{ count: number; scrip_codes: any[] }> {
    return this.http.get<{ count: number; scrip_codes: any[] }>(`${this.apiUrl}/bse/scrip-codes`);
  }
}

File: mobile/lib/core/services/announcement_service.dart
import '../models/announcement.dart';
import 'api_service.dart';
import '../../config/api_config.dart';

class AnnouncementService {
  // Get NSE announcements for a symbol
  static Future<List<NSEAnnouncement>> getNSEAnnouncements(
    String symbol, {
    String? fromDate,
    String? toDate,
    int limit = 100,
  }) async {
    final params = <String, String>{'limit': limit.toString()};
    if (fromDate != null) params['from_date'] = fromDate;
    if (toDate != null) params['to_date'] = toDate;

    final response = await ApiService.get(
      ApiConfig.nseAnnouncements(symbol),
      queryParams: params,
    );
    final data = ApiService.decodeResponse(response);
    return (data['announcements'] as List? ?? [])
        .map((e) => NSEAnnouncement.fromJson(e))
        .toList();
  }

  // Get BSE announcements
  static Future<Map<String, dynamic>> getBSEAnnouncements({
    String? scripCode,
    String? fromDate,
    String? toDate,
    int page = 1,
  }) async {
    final params = <String, String>{'page': page.toString()};
    if (scripCode != null) params['scrip_code'] = scripCode;
    if (fromDate != null) params['from_date'] = fromDate;
    if (toDate != null) params['to_date'] = toDate;

    final response = await ApiService.get(
      ApiConfig.bseAnnouncements,
      queryParams: params,
    );
    final data = ApiService.decodeResponse(response);
    final announcements = (data['announcements'] as List? ?? [])
        .map((e) => BSEAnnouncement.fromJson(e))
        .toList();
    return {
      'announcements': announcements,
      'total_pages': data['total_pages'] ?? 0,
      'current_page': data['current_page'] ?? 1,
    };
  }

  // NSE symbol autocomplete
  static Future<List<Map<String, String>>> nseAutocomplete(String query) async {
    if (query.isEmpty) return [];
    final response = await ApiService.get(
      ApiConfig.nseAutocomplete,
      queryParams: {'q': query},
    );
    final data = ApiService.decodeResponse(response);
    final results = data['results'] as List? ?? [];
    return results
        .map<Map<String, String>>(
          (e) => {
            'symbol': e['symbol']?.toString() ?? '',
            'company_name': e['company_name']?.toString() ?? '',
          },
        )
        .toList();
  }
}
